cirq/linalg/transformations_test.py

```diff
@@ -393,3 +394,203 @@ def test_partial_trace_invalid_inputs():
+
+
+def test_subwavefunction_invalid_inputs():
+
+    # State cannot be expressed as a qubit wavefunction.
+    with pytest.raises(ValueError, match='7'):
+        cirq.subwavefunction(np.arange(7), [1, 2], atol=1e-8)
+
+    # State shape does not conform to input requirements.
+    with pytest.raises(ValueError, match='shaped'):
+        cirq.subwavefunction(np.arange(16).reshape(2, 4, 2), [1, 2], atol=1e-8)
+    with pytest.raises(ValueError, match='shaped'):
+        cirq.subwavefunction(np.arange(16).reshape((16, 1)), [1, 2], atol=1e-8)
+
+    with pytest.raises(ValueError, match='normalized'):
+        cirq.subwavefunction(np.arange(16), [1, 2], atol=1e-8)
+
+    # Bad choice of input indices.
+    with pytest.raises(ValueError, match='2, 2'):
+        cirq.subwavefunction(
+            np.arange(16) / np.linalg.norm(np.arange(16)),
+            [1, 2, 2],
+            atol=1e-8)
+    with pytest.raises(ValueError, match='invalid'):
+        cirq.subwavefunction(
+            np.array([1, 0, 0, 0]).reshape(2, 2), [5], atol=1e-8)
+    with pytest.raises(ValueError, match='invalid'):
+        cirq.subwavefunction(
+            np.array([1, 0, 0, 0]).reshape(2, 2), [0, 1, 2], atol=1e-8)
+
+
+def test_wavefunction_partial_trace_invalid_input():
+    # TODO
+    pass
+
+
+def mixtures_equal(m1, m2, atol=1e-7):
+    if len(m1) != len(m2):
+        return False
+    for (p1, v1), (p2, v2) in zip(m1, m2):
+        if not (cirq.approx_eq(p1, p2, atol=atol)
+                and cirq.equal_up_to_global_phase(v1, v2, atol=atol)):
+            return False
+    return True
+
+
+def test_wavefunction_partial_trace_pure_result():
+    a = cirq.testing.random_superposition(4)
+    b = cirq.testing.random_superposition(8)
+    c = cirq.testing.random_superposition(16)
+    state = np.kron(np.kron(a, b), c).reshape((2,) * 9)
+
+    assert mixtures_equal(
+        cirq.wavefunction_partial_trace(state, [0, 1], atol=1e-8),
+        ((1.0, a.reshape(2, 2)),))
+    assert mixtures_equal(
+        cirq.wavefunction_partial_trace(state, [2, 3, 4], atol=1e-8),
+        ((1.0, b.reshape(2, 2, 2)),))
+    assert mixtures_equal(
+        cirq.wavefunction_partial_trace(state, [5, 6, 7, 8], atol=1e-8),
+        ((1.0, c.reshape(2, 2, 2, 2)),))
+    assert mixtures_equal(
+        cirq.wavefunction_partial_trace(state, [0, 1, 2, 3, 4], atol=1e-8),
+        ((1.0, np.kron(a, b).reshape(2, 2, 2, 2, 2)),))
+    assert mixtures_equal(
+        cirq.wavefunction_partial_trace(state, [0, 1, 5, 6, 7, 8], atol=1e-8),
+        ((1.0, np.kron(a, c).reshape(2, 2, 2, 2, 2, 2)),))
+    assert mixtures_equal(
+        cirq.wavefunction_partial_trace(
+            state, [2, 3, 4, 5, 6, 7, 8], atol=1e-8),
+        ((1.0, np.kron(b, c).reshape(2, 2, 2, 2, 2, 2, 2)),))
+
+    # Return mixture will defer to numpy.linalg.eig's builtin tolerance.
+    assert mixtures_equal(
+        cirq.wavefunction_partial_trace(state, [0, 1], atol=1e-20),
```

Comment:
You're using a strict tolerance in the call, but not in the check.

Reply:
fixed.